qualcommax: ipq60xx: add TP-Link EAP625-Outdoor HD v1 support

mrnuke opened the pull request:
This board as a few minor TODO items, but is otherwise ready to go.
- [x] BDF pull request: https://github.com/openwrt/firmware_qca-wireless/pull/87
- [x] ath11k driver does not detect regdb in board data file (fixed in DBF PR)

There are further cleanups that are beyond the scope of this PR:
- Bluetooth support
- Unify devicetree with EAP623-Outdoor


The commit message
-------------------------------

TP-Link EAP625-Outdoor HD is a 802.11ax AP claiming AX1800 support.
It is wall or pole mountable, and rated for outdoor use. It can only
be powered via PoE.

Hardware-wise, it is very similar to the older EAP610-Outdoor model.
Software-wise, I couldn't find a difference other than the board
data files, and device name. For this reason, the majority of the
devicetree from the EAP610-Outdoor is reused.

This device currently comes in a "v1", and "v1.6" version. The
"support-list" of the vendor firmware does not have a distinction
between these versions. This commit was tested on a 'V1.6" device.

Specifications:
---------------
* CPU: Qualcomm IPQ6018 Quad core Cortex-A53
* RAM: 512 MB
* Storage: 128MB NAND
* Ethernet:
  * Gigabit RJ45 port with PoE input
* WLAN:
  * 2.4GHz/5GHz
* LEDs:
  * Multi-color System LED (Green/Amber)
* Buttons:
  * 1x Reset
* UART: 4-pin unpopulated header
  * 1.8 V level, Pinout 1 - TX, 2 - RX, 3 - GND, 4 - 1.8V

Installation:
=============

Web UI method
-------------

Set up the device using the vendor's web UI. After that go to
Management->SSH and enable the "SSH Login" checkbox. Select "Save".
The connect to the machine via SSH:

ssh -o hostkeyalgorithms=ssh-rsa <ip_of_device>

Disable signature verification:

cliclientd stopcs

Rename the "-web-ui-factory" image to something less than 63
characters, maintaining the ".bin" suffix.
 * Go to System -> Firmware Update.
 * Under "New Firmware File", click "Browse" and select the image
 * Select "Update" and confirm by clicking "OK".

If the update fails, the web UI should show an error message.
Otherwise, the device should reboot into OpenWRT.

NOTE: If ssh continues to complain that "no matching host key type
found. Their offer: ssh-rsa,ssh-dss" it likely means that yor distro
has completely disabled deprecated siphers in ssh. In that case, run
the ssh command from a docker container of an older distro.

TFTP method
-----------

To flash via tftp, first place the initramfs image on the TFTP server.

setenv serverip <ip of tftp server>
setenv ipaddr <ip in same subnet as tftp server>
tftpboot tplink_eap625-outdoor-hd-v1-initramfs-uImage.itb
bootm

This should boot OpenWRT. Once booted, flash the sysupgrade.bin image
using either luci or the commandline.

mrnuke committed:
qualcommax: ipq60xx: split eap610-outdoor devicetree

I have an EAP625-Outdoor HD v1 that is very similar with the EAP610
Outdoor. It works with the EAP610 initramfs, without any obvious loss
of functionality. In order to prepare for supporting the EAP625, move
the common parts into a shared .dtsi.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18584
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommax: ipq60xx: add TP-Link EAP625-Outdoor HD v1 support

TP-Link EAP625-Outdoor HD is a 802.11ax AP claiming AX1800 support.
It is wall or pole mountable, and rated for outdoor use. It can only
be powered via PoE.

Hardware-wise, it is very similar to the older EAP610-Outdoor model.
Software-wise, I couldn't find a difference other than the board
data files, and device name. For this reason, the majority of the
devicetree from the EAP610-Outdoor is reused.

This device currently comes in a "v1", and "v1.6" version. The
"support-list" of the vendor firmware does not have a distinction
between these versions. This commit was tested on a 'V1.6" device.

Specifications:
---------------
* CPU: Qualcomm IPQ6018 Quad core Cortex-A53
* RAM: 512 MB
* Storage: 128MB NAND
* Ethernet:
  * Gigabit RJ45 port with PoE input
* WLAN:
  * 2.4GHz/5GHz
* LEDs:
  * Multi-color System LED (Green/Amber)
* Buttons:
  * 1x Reset
* UART: 4-pin unpopulated header
  * 1.8 V level, Pinout 1 - TX, 2 - RX, 3 - GND, 4 - 1.8V

Installation:
=============

Web UI method
-------------

Set up the device using the vendor's web UI. After that go to
Management->SSH and enable the "SSH Login" checkbox. Select "Save".
The connect to the machine via SSH:

ssh -o hostkeyalgorithms=ssh-rsa <ip_of_device>

Disable signature verification:

cliclientd stopcs

Rename the "-web-ui-factory" image to something less than 63
characters, maintaining the ".bin" suffix.
 * Go to System -> Firmware Update.
 * Under "New Firmware File", click "Browse" and select the image
 * Select "Update" and confirm by clicking "OK".

If the update fails, the web UI should show an error message.
Otherwise, the device should reboot into OpenWRT.

NOTE: If ssh continues to complain that "no matching host key type
found. Their offer: ssh-rsa,ssh-dss" it likely means that yor distro
has completely disabled deprecated siphers in ssh. In that case, run
the ssh command from a docker container of an older distro.

TFTP method
-----------

To flash via tftp, first place the initramfs image on the TFTP server.

setenv serverip <ip of tftp server>
setenv ipaddr <ip in same subnet as tftp server>
tftpboot tplink_eap625-outdoor-hd-v1-initramfs-uImage.itb
bootm

This should boot OpenWRT. Once booted, flash the sysupgrade.bin image
using either luci or the commandline.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18584
Signed-off-by: Robert Marko <robimarko@gmail.com>

185264646 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-eap610-outdoor.dts:
Please also do the same thing for eap623 outdoor.

mrnuke replied:
This suggestion makes perfect sense, and is on my radar. In order to keep the review short, I have elected not to touch the EAP623 in this PR.

185264646 replied:
> This suggestion makes perfect sense, and is on my radar. In order to keep the review short, I have elected not to touch the EAP623 in this PR.

According to the GPL code released by TP-Link. EAP623 and EAP625 are highly identical. So i guess only minor modification is needed to get EAP625 support based on EAP623.

185264646 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-tplink-eap6xx-outdoor.dtsi:
```diff
@@ -0,0 +1,146 @@
+
+&switch {
+	switch_lan_bmp = <ESS_PORT5>;
+	switch_mac_mode1 = <MAC_MODE_SGMII_CHANNEL0>;
```
<MAC_MODE_SGMII_PLUS> or <MAC_MODE_SGMII>.

robimarko replied:
If its 2.5G then `MAC_MODE_SGMII_PLUS` for sure

robimarko replied:
I know, but vendor FW-s just default to it basically, for 1G MAC_MODE_SGMII_CHANNEL0 is fine

mrnuke replied:
The PHY ID is 0x001cc916 on both 610 and 625 devices. This is a "RTL8211F Gigabit Ethernet". Thus 1G.

185264646 commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -183,6 +183,26 @@ define Device/tplink_eap623od-hd-v1
+		EAP625-Outdoor HD(TP-Link|KR|AX1800-D):1.0
+
+endef
+TARGET_DEVICES += tplink_eap625-outdoor-hd-v1
```
Inconsistant with eap623 outdoor. Please rename either one.

mrnuke replied:
I will rework the eap623-outdoor naming and devicetree unification in a separate PR. What do you think?

185264646 commented on target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network:
```diff
@@ -33,7 +33,8 @@ ipq60xx_setup_interfaces()
 	tplink,eap610-outdoor|\
-	tplink,eap623od-hd-v1)
+	tplink,eap623od-hd-v1|\
+	tplink,eap625-outdoor-hd-v1)
```
same above.

185264646 commented:
Seems like EAP625 comes with BLE support. Have you tried get that working?

mrnuke commented:
> Seems like EAP625 comes with BLE support. Have you tried get that working?

As I tried to explain at the top of this PR's description I am aware of the bluetooth. I have chosen not to focus on it in this review.

185264646 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-tplink-eap6xx-outdoor.dtsi:
```diff
@@ -0,0 +1,146 @@
+	reset-post-delay-us = <50000>;
+	status = "okay";
+
+	rtl8211f_4: ethernet-phy@4 {
```
please add a pinctrl node for this. see https://github.com/openwrt/openwrt/pull/18450/files#diff-ffc75b8a82a6b801f50d3aafbea208b7e6d637d474f598781906479121091495R92 .

mrnuke replied:
Done.

mrnuke committed:
qualcommax: ipq60xx: eap6xx-outdoor: add PHY reset pinctrl

It was suggested that the pinctrl property of the mdio node should
include the PHY reset GPIO. Add them in.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18584
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Thanks! Rebased on top of main and merged!
